commit 851ae6424697d1c4f085cb878c88168923ebcad1
Author: Ying Huang <huang.ying.caritas@gmail.com>
Date:   Mon Apr 17 07:59:29 2023 +0800

    migrate_pages_batch: fix statistics for longterm pin retry
    
    In commit fd4a7ac32918 ("mm: migrate: try again if THP split is failed due
    to page refcnt"), if the THP splitting fails due to page reference count,
    we will retry to improve migration successful rate.  But the failed
    splitting is counted as migration failure and migration retry, which will
    cause duplicated failure counting.  So, in this patch, this is fixed via
    undoing the failure counting if we decide to retry.  The patch is tested
    via failure injection.
    
    Link: https://lkml.kernel.org/r/20230416235929.1040194-1-ying.huang@intel.com
    Fixes: fd4a7ac32918 ("mm: migrate: try again if THP split is failed due to page refcnt")
    Signed-off-by: "Huang, Ying" <ying.huang@intel.com>
    Reviewed-by: Baolin Wang <baolin.wang@linux.alibaba.com>
    Cc: Alistair Popple <apopple@nvidia.com>
    Cc: David Hildenbrand <david@redhat.com>
    Cc: Yang Shi <shy828301@gmail.com>
    Cc: Zi Yan <ziy@nvidia.com>
    Signed-off-by: Andrew Morton <akpm@linux-foundation.org>

diff --git a/mm/migrate.c b/mm/migrate.c
index 41f2154d66ad..5d95e09b1618 100644
--- a/mm/migrate.c
+++ b/mm/migrate.c
@@ -1710,6 +1710,9 @@ static int migrate_pages_batch(struct list_head *from, new_page_t get_new_page,
 							large_retry++;
 							thp_retry += is_thp;
 							nr_retry_pages += nr_pages;
+							/* Undo duplicated failure counting. */
+							nr_large_failed--;
+							stats->nr_thp_failed -= is_thp;
 							break;
 						}
 					}